extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Descriptor.scala

```diff
@@ -0,0 +1,184 @@
+import org.apache.kyuubi.plugin.spark.authz.serde.TableExtractor.tableExtractors
+import org.apache.kyuubi.plugin.spark.authz.serde.TableType.TableType
+import org.apache.kyuubi.plugin.spark.authz.serde.TableTypeExtractor.tableTypeExtractors
+import org.apache.kyuubi.plugin.spark.authz.util.AuthZUtils._
+
+/**
+ * A database object(such as database, table, function) descriptor describes its name and getter
+ * in/from another object(such as a spark sql command).
+ */
+sealed trait Descriptor {
+
+  /**
+   * Describes the field name, such as a table field name in the spark logical plan.
+   *
+   * @return the database object field name
+   */
+  def fieldName: String
+
+  /**
+   * The extractor/getter classname for the field
+   */
+  def fieldExtractor: String
+
+  def getValue(v: AnyRef): AnyRef
+
+  final def error(v: AnyRef, e: Throwable): String = {
+    val resourceName = getClass.getSimpleName.stripSuffix("Desc")
+    val objectClass = v.getClass.getName
+    s"[Spark$SPARK_VERSION] failed to get $resourceName from $objectClass by" +
+      s" $fieldExtractor/$fieldName, due to ${e.getMessage}"
+  }
+}
+
+case class ColumnDesc(
+    fieldName: String,
+    fieldExtractor: String) extends Descriptor {
+
+  override def getValue(v: AnyRef): Seq[String] = {
+    val columnsVal = invoke(v, fieldName)
+    val columnExtractor = columnExtractors(fieldExtractor)
+    columnExtractor(columnsVal)
+  }
+}
+
+case class DatabaseDesc(
+    fieldName: String,
+    fieldExtractor: String,
+    isInput: Boolean = false) extends Descriptor {
+  override def getValue(v: AnyRef): String = {
+    val databaseVal = invoke(v, fieldName)
+    val databaseExtractor = dbExtractors(fieldExtractor)
+    databaseExtractor(databaseVal)
+  }
+}
+
+case class FunctionTypeDesc(
+    fieldName: String,
+    fieldExtractor: String,
+    skipTypes: Seq[String]) extends Descriptor {
+
+  override def getValue(v: AnyRef): FunctionType = {
+    getValue(v, SparkSession.active)
+  }
+
+  def getValue(v: AnyRef, spark: SparkSession): FunctionType = {
+    val functionTypeVal = invoke(v, fieldName)
+    val functionTypeExtractor = functionTypeExtractors(fieldExtractor)
+    functionTypeExtractor(functionTypeVal, spark)
+  }
+
+  def skip(v: AnyRef, spark: SparkSession): Boolean = {
+    skipTypes.exists(skipType => getValue(v, spark) == FunctionType.withName(skipType))
+  }
+}
+
+case class FunctionDesc(
+    fieldName: String,
+    fieldExtractor: String,
+    databaseDesc: Option[DatabaseDesc] = None,
+    functionTypeDesc: Option[FunctionTypeDesc] = None,
+    isInput: Boolean = false) extends Descriptor {
+  override def getValue(v: AnyRef): Function = {
+    val functionVal = invoke(v, fieldName)
+    val functionExtractor = functionExtractors(fieldExtractor)
+    var function = functionExtractor(functionVal)
+    if (function.database.isEmpty) {
+      function = function.copy(database = databaseDesc.map(_.getValue(v)))
```

Comment:
why `TableDesc. getValue` does not need fill database ?

Reply:
I think `TableDesc.getValue` is filling database of `Table` in `TableExtractor` implementations.
Is this a question for FunctionDesc or else?